File: Client/orgTable/code/ProtocolSkill.cs
using System;
using System.Text;

namespace Protocol
{
	public class Skill : Protocol.IProtocolStream
	{
		public UInt16 id;
		public byte level;

		#region METHOD

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint16(buffer, ref pos_, id);
				BaseDLL.encode_int8(buffer, ref pos_, level);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint16(buffer, ref pos_, ref id);
				BaseDLL.decode_int8(buffer, ref pos_, ref level);
			}
		#endregion

	}

	public class ChangeSkill : Protocol.IProtocolStream
	{
		public UInt16 id;
		public byte dif;

		#region METHOD

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint16(buffer, ref pos_, id);
				BaseDLL.encode_int8(buffer, ref pos_, dif);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint16(buffer, ref pos_, ref id);
				BaseDLL.decode_int8(buffer, ref pos_, ref dif);
			}
		#endregion

	}

	public class Buff : Protocol.IProtocolStream
	{
		public UInt64 uid;
		public UInt32 id;
		public UInt32 overlay;
		public UInt32 duration;

		#region METHOD

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint64(buffer, ref pos_, uid);
				BaseDLL.encode_uint32(buffer, ref pos_, id);
				BaseDLL.encode_uint32(buffer, ref pos_, overlay);
				BaseDLL.encode_uint32(buffer, ref pos_, duration);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint64(buffer, ref pos_, ref uid);
				BaseDLL.decode_uint32(buffer, ref pos_, ref id);
				BaseDLL.decode_uint32(buffer, ref pos_, ref overlay);
				BaseDLL.decode_uint32(buffer, ref pos_, ref duration);
			}
		#endregion

	}

	[Protocol]
	public class SceneChangeSkillsReq : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500701;
		public ChangeSkill[] skills = new ChangeSkill[0];

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint16(buffer, ref pos_, (UInt16)skills.Length);
				for(int i = 0; i < skills.Length; i++)
				{
					skills[i].encode(buffer, ref pos_);
				}
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				UInt16 skillsCnt = 0;
				BaseDLL.decode_uint16(buffer, ref pos_, ref skillsCnt);
				skills = new ChangeSkill[skillsCnt];
				for(int i = 0; i < skills.Length; i++)
				{
					skills[i] = new ChangeSkill();
					skills[i].decode(buffer, ref pos_);
				}
			}
		#endregion

	}

	[Protocol]
	public class SceneChangeSkillsRes : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500702;
		public UInt32 result;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, result);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref result);
			}
		#endregion

	}

	[Protocol]
	public class SceneAddBuff : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500711;
		public UInt32 buffId;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, buffId);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref buffId);
			}
		#endregion

	}

	[Protocol]
	public class SceneNotifyRemoveBuff : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500712;
		public UInt32 buffId;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, buffId);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref buffId);
			}
		#endregion

	}

}

File: Client/orgTable/code/ProtocolSkill.lua
local descriptor = require "descriptor"

module "ProtocolSkill"

Skill = descriptor.def_struct("Skill", 
	{
		descriptor.def_scalar_field("id", descriptor.type_uint16, 0),
		descriptor.def_scalar_field("level", descriptor.type_uint8, 0),
	}
)

ChangeSkill = descriptor.def_struct("ChangeSkill", 
	{
		descriptor.def_scalar_field("id", descriptor.type_uint16, 0),
		descriptor.def_scalar_field("dif", descriptor.type_uint8, 0),
	}
)

Buff = descriptor.def_struct("Buff", 
	{
		descriptor.def_scalar_field("uid", descriptor.type_uint64, 0),
		descriptor.def_scalar_field("id", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("overlay", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("duration", descriptor.type_uint32, 0),
	}
)

SceneChangeSkillsReq = descriptor.def_message("SceneChangeSkillsReq", 500701, 
	{
		descriptor.def_message_vector_field("skills", ChangeSkill),
	}
)

SceneChangeSkillsRes = descriptor.def_message("SceneChangeSkillsRes", 500702, 
	{
		descriptor.def_scalar_field("result", descriptor.type_uint32, 0),
	}
)

SceneAddBuff = descriptor.def_message("SceneAddBuff", 500711, 
	{
		descriptor.def_scalar_field("buffId", descriptor.type_uint32, 0),
	}
)

SceneNotifyRemoveBuff = descriptor.def_message("SceneNotifyRemoveBuff", 500712, 
	{
		descriptor.def_scalar_field("buffId", descriptor.type_uint32, 0),
	}
)

